numpy/__init__.pyi

```diff
@@ -48,15 +48,77 @@ else:
 def __getattr__(name: str) -> Any: ...
 
 _NdArraySubClass = TypeVar("_NdArraySubClass", bound=ndarray)
+_DTypeScalar = TypeVar("_DTypeScalar", bound=generic)
 
-class dtype:
+class dtype(Generic[_DTypeScalar]):
     names: Optional[Tuple[str, ...]]
-    def __init__(
-        self,
-        dtype: DtypeLike,
+    # Overload for subclass of generic
+    @overload
+    def __new__(
+        cls,
+        dtype: Type[_DTypeScalar],
         align: bool = ...,
         copy: bool = ...,
-    ) -> None: ...
+    ) -> dtype[_DTypeScalar]: ...
+    # Overloads for string aliases
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal["float64"],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[float64]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal["float32"],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[float32]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal["int64"],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[int64]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Literal["int32"],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[int32]: ...
+    # Overloads for Python types. Order is important here.
+    @overload
+    def __new__(
+        cls,
+        dtype: Type[bool],
+        align: bool = ...,
+        copy: bool = ...,
+    ) -> dtype[bool_]: ...
+    @overload
+    def __new__(
+        cls,
+        dtype: Type[int],
```

Comment:
We'll have to be careful with `int` and `float` here as the corresponding a `np.generic` is platform dependent.
This is particularly true for Windows as it generally (always?) uses 32-bit integers, 
both on 32- and 64-bit operating systems.

Reply:
Ah yeah, I think we want `np.intc`, but we haven't added an annotation for it yet!

Reply:
> Ah yeah, I think we want `np.intc`, but we haven't added an annotation for it yet!

On MacOS it seems that `np.intc is np.int32` while `np.dtype(int) == np.int64`.
I believe `np.int_` does to the trick though.

In any case, it appears that the likes of mypy support platform checks ([ref](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks)), which might be exactly what we need here!

Reply:
`np.dtype(int)` and `int_` are synonyms on all platforms.

Reply:
> `np.dtype(int)` and `np.dtype(np.int_)` are synonyms on all platforms.

Good to know, then this is exactly what we're looking for! (considering this is about adding a `np.dtype(int)` overload.

I think the code below should trick unless there are more situations where `np.int_ is np.int32`.

```python
if sys.platform == "win32":
    int_ = np.int32  # Windows-specific code
else:
    int_ = np.int64  # Other systems

class dtype:
    def __new__(cls, dtype: Type[int], ...) -> dtype[int_]: ...

```

Reply:
Are `if`s legal like that in subfiles? If so, is `sys.platform` whitelisted as one of the things you can access?

Reply:
Most 32-bit platforms will likely have `int_ == int32` too  - windows is special, because it does this even on 64-bit (so as to be consistent with 32-bit which came first)

Reply:
> Are `ifs` legal like that in subfiles? If so, is `sys.platform` whitelisted as one of the things you can access?

A select few of them are ([ref](https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks)).
This includes `sys.version_info` (which we've used previously) and apparently also `sys.platform`, yes.

Reply:
I think the mostly correct logic is:
```
if sys.maxsize > 2**32:
    intp = np.int64
else:
    intp = np.int32

if sys.platform == "win32":
    int_ = np.int32
else:
    int_ = np.intp
```
You might need to make an upstream patch to mypy to support `sys.maxsize`.

In the meantime, I suppose you can define `intp = Union[np.int32, np.int64]`?

Reply:
We could also potentially fake `np.intp` and `np.int_` as being their own types. If one were writing code that needed to take variation across systems into account then the type checker complaining about something like `intp + int64` by default regardless of platform might be the most helpful.

Reply:
Yet another option would be to add some sort of dynamically generated `.pyi` file (perhaps whenever `setup.py` is runned?) for handling all these platform-specific `generic` aliases and then import them from there. 
Considering there is quite a number of them ([ref](https://numpy.org/devdocs/user/basics.types.html)).

Reply:
The dynamically generated `.pyi` file sounds like a nice solution long term-for now I've demoted those returns to `dtype[Any]` so as to not get blocked here on solving the platform-dependent dtype problem.